src/test/shared/sam/debugger/samDebugConfigProvider.test.ts

```diff
@@ -273,6 +273,29 @@ describe('SamDebugConfigurationProvider', async function () {
     })
 
     describe('makeConfig', async function () {
+        describe('buildDir', function () {
+            it('uses `buildDir` as `baseBuildDir` when provided', async function () {
+                const buildDir = '/my/build/dir'
+                const folder = testutil.getWorkspaceFolder(testutil.getProjectDir())
+                const launchConfig = await getConfig(
+                    debugConfigProvider,
+                    ext.templateRegistry,
+                    'testFixtures/workspaceFolder/js-plain-sam-app/'
+                )
+                const config = launchConfig.config as AwsSamDebuggerConfiguration & {
+                    invokeTarget: { target: 'template' }
+                }
+                config.sam = Object.assign(launchConfig.config.sam ?? {}, { buildDir })
+                config.invokeTarget.templatePath = config.invokeTarget.templatePath.replace(
+                    '${workspaceFolder}',
+                    launchConfig.folder.uri.fsPath
+                )
+
+                const actual = await debugConfigProvider.makeConfig(folder, config)
+                assert.strictEqual(actual?.baseBuildDir, buildDir)
```

Comment:
this only tests a very narrow part of the configuration. Could add buildDir to the tests below instead or in addition.

Otherwise, we aren't testing how all of the configuration state affects each other. That results in nice-looking tests, but covers much less.

Reply:
That's kind of why I'm saying we should separate/isolate various parts of the state. By doing that, we can enumerate through all possible permutations of state with tidy tests.